### ISSUE TITLE
Assertion `isa<X>(Val) && "cast<Ty>() argument of incompatible type!"' failed.

### ISSUE THREAD

**Issue by nmeum** [Microsoft/checkedc-clang#436]
I am currently playing around with Checked C and while doing so I believe I found a bug in the compiler that hasn't been reported so far. I tried to reduce the code initially triggering the problem to a minimal example and came up with the code below.

I used the following version of the Checked C compiler:

```
$ clang --version
clang version 6.0.0 (https://github.com/Microsoft/checkedc-clang 48bd8fda1e5d1d929a64e9ad7b932bc6d9399591) (https://github.com/Microsoft/checkedc-llvm 9bf0f4a1bf12d8f7bd6259309b9352e3e93a396b)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/checkedc-llvm/bin
```

And when I attempted to compile the following program:

```C
#include <stdint.h>
#include <sys/types.h>

static uint8_t *test : itype(_Ptr<uint8_t>);

int testfn(uint8_t *buf : count(len), size_t len)
_Checked {
	test = buf;
	return 0;
}
```

clang emitted this error message:

```
$ /opt/checkedc-llvm/bin/clang -fcheckedc-extension -v -c -o checkedc_bug.o checkedc_bug.c
checkedc_bug.c:8:9: warning: cannot prove cast source bounds are wide enough for '_Ptr<uint8_t>' (aka '_Ptr<unsigned char>') [-Wcheck-bounds-decls]
        test = buf;
               ^~~
checkedc_bug.c:8:9: note: (expanded) required bounds are 'bounds(buf, buf + 1)'
checkedc_bug.c:8:9: note: (expanded) inferred bounds are 'bounds(buf, buf + len)'
clang-5.0: /root/checkedc/llvm/include/llvm/Support/Casting.h:255: typename llvm::cast_retty<X, Y*>::ret_type llvm::cast(Y*) [with X = clang::ExplicitCastExpr; Y = const clang::CastExpr; typename llvm::cast_retty<X, Y*>::ret_type = const clang::ExplicitCastExpr*]: Assertion `isa<X>(Val) && "cast<Ty>() argument of incompatible type!"' failed.
#0 0x0000561fb76c6924 llvm::sys::PrintStackTrace(llvm::raw_ostream&) /root/checkedc/llvm/lib/Support/Unix/Signals.inc:398:0
#1 0x0000561fb76c69b7 PrintStackTraceSignalHandler(void*) /root/checkedc/llvm/lib/Support/Unix/Signals.inc:462:0
#2 0x0000561fb76c4e62 llvm::sys::RunSignalHandlers() /root/checkedc/llvm/lib/Support/Signals.cpp:49:0
#3 0x0000561fb76c6299 SignalHandler(int) /root/checkedc/llvm/lib/Support/Unix/Signals.inc:252:0
#4 0x00007f4dcef3c3b0 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x123b0)
#5 0x00007f4dcdcdba70 gsignal (/lib/x86_64-linux-gnu/libc.so.6+0x33a70)
#6 0x00007f4dcdcdd19a abort (/lib/x86_64-linux-gnu/libc.so.6+0x3519a)
#7 0x00007f4dcdcd4667 (/lib/x86_64-linux-gnu/libc.so.6+0x2c667)
#8 0x00007f4dcdcd4712 (/lib/x86_64-linux-gnu/libc.so.6+0x2c712)
#9 0x0000561fb7ccae40 llvm::cast_retty<clang::ExplicitCastExpr, clang::CastExpr const*>::ret_type llvm::cast<clang::ExplicitCastExpr, clang::CastExpr const>(clang::CastExpr const*) /root/checkedc/llvm/include/llvm/Support/Casting.h:255:0
#10 0x0000561fb7cc332f clang::CodeGen::CodeGenFunction::EmitCastLValue(clang::CastExpr const*) /root/checkedc/llvm/tools/clang/lib/CodeGen/CGExpr.cpp:4114:0
#11 0x0000561fb7cb1f9e clang::CodeGen::CodeGenFunction::EmitLValue(clang::Expr const*) /root/checkedc/llvm/tools/clang/lib/CodeGen/CGExpr.cpp:1243:0
#12 0x0000561fb7cb1337 clang::CodeGen::CodeGenFunction::EmitCheckedLValue(clang::Expr const*, clang::CodeGen::CodeGenFunction::TypeCheckKind) /root/checkedc/llvm/tools/clang/lib/CodeGen/CGExpr.cpp:1093:0
#13 0x0000561fb7cf38b7 (anonymous namespace)::ScalarExprEmitter::EmitCheckedLValue(clang::Expr const*, clang::CodeGen::CodeGenFunction::TypeCheckKind) /root/checkedc/llvm/tools/clang/lib/CodeGen/CGExprScalar.cpp:240:0
#14 0x0000561fb7d036b2 (anonymous namespace)::ScalarExprEmitter::VisitBinAssign(clang::BinaryOperator const*) /root/checkedc/llvm/tools/clang/lib/CodeGen/CGExprScalar.cpp:3336:0
#15 0x0000561fb7d07706 clang::StmtVisitorBase<clang::make_ptr, (anonymous namespace)::ScalarExprEmitter, llvm::Value*>::Visit(clang::Stmt*) /root/checkedc/llvm/tools/clang/include/clang/AST/StmtVisitor.h:71:0
#16 0x0000561fb7cf3d8a (anonymous namespace)::ScalarExprEmitter::Visit(clang::Expr*) /root/checkedc/llvm/tools/clang/lib/CodeGen/CGExprScalar.cpp:358:0
#17 0x0000561fb7d060aa clang::CodeGen::CodeGenFunction::EmitScalarExpr(clang::Expr const*, bool) /root/checkedc/llvm/tools/clang/lib/CodeGen/CGExprScalar.cpp:3851:0
#18 0x0000561fb7cac8d0 clang::CodeGen::CodeGenFunction::EmitAnyExpr(clang::Expr const*, clang::CodeGen::AggValueSlot, bool) /root/checkedc/llvm/tools/clang/lib/CodeGen/CGExpr.cpp:178:0
#19 0x0000561fb7cac858 clang::CodeGen::CodeGenFunction::EmitIgnoredExpr(clang::Expr const*) /root/checkedc/llvm/tools/clang/lib/CodeGen/CGExpr.cpp:163:0
#20 0x0000561fb79e3181 clang::CodeGen::CodeGenFunction::EmitStmt(clang::Stmt const*, llvm::ArrayRef<clang::Attr const*>) /root/checkedc/llvm/tools/clang/lib/CodeGen/CGStmt.cpp:109:0
#21 0x0000561fb79e3f14 clang::CodeGen::CodeGenFunction::EmitCompoundStmtWithoutScope(clang::CompoundStmt const&, bool, clang::CodeGen::AggValueSlot) /root/checkedc/llvm/tools/clang/lib/CodeGen/CGStmt.cpp:379:0
#22 0x0000561fb7a53782 clang::CodeGen::CodeGenFunction::EmitFunctionBody(clang::CodeGen::FunctionArgList&, clang::Stmt const*) /root/checkedc/llvm/tools/clang/lib/CodeGen/CodeGenFunction.cpp:1138:0
#23 0x0000561fb7a54300 clang::CodeGen::CodeGenFunction::GenerateCode(clang::GlobalDecl, llvm::Function*, clang::CodeGen::CGFunctionInfo const&) /root/checkedc/llvm/tools/clang/lib/CodeGen/CodeGenFunction.cpp:1303:0
#24 0x0000561fb7a7254e clang::CodeGen::CodeGenModule::EmitGlobalFunctionDefinition(clang::GlobalDecl, llvm::GlobalValue*) /root/checkedc/llvm/tools/clang/lib/CodeGen/CodeGenModule.cpp:3261:0
#25 0x0000561fb7a6dd8e clang::CodeGen::CodeGenModule::EmitGlobalDefinition(clang::GlobalDecl, llvm::GlobalValue*) /root/checkedc/llvm/tools/clang/lib/CodeGen/CodeGenModule.cpp:2039:0
#26 0x0000561fb7a6cf0b clang::CodeGen::CodeGenModule::EmitGlobal(clang::GlobalDecl) /root/checkedc/llvm/tools/clang/lib/CodeGen/CodeGenModule.cpp:1816:0
#27 0x0000561fb7a76018 clang::CodeGen::CodeGenModule::EmitTopLevelDecl(clang::Decl*) /root/checkedc/llvm/tools/clang/lib/CodeGen/CodeGenModule.cpp:4003:0
#28 0x0000561fb848d9cc (anonymous namespace)::CodeGeneratorImpl::HandleTopLevelDecl(clang::DeclGroupRef) /root/checkedc/llvm/tools/clang/lib/CodeGen/ModuleBuilder.cpp:159:0
#29 0x0000561fb8487699 clang::BackendConsumer::HandleTopLevelDecl(clang::DeclGroupRef) /root/checkedc/llvm/tools/clang/lib/CodeGen/CodeGenAction.cpp:170:0
#30 0x0000561fb922eabf clang::ParseAST(clang::Sema&, bool, bool) /root/checkedc/llvm/tools/clang/lib/Parse/ParseAST.cpp:151:0
#31 0x0000561fb7fa4769 clang::ASTFrontendAction::ExecuteAction() /root/checkedc/llvm/tools/clang/lib/Frontend/FrontendAction.cpp:998:0
#32 0x0000561fb848623c clang::CodeGenAction::ExecuteAction() /root/checkedc/llvm/tools/clang/lib/CodeGen/CodeGenAction.cpp:1032:0
#33 0x0000561fb7fa41c3 clang::FrontendAction::Execute() /root/checkedc/llvm/tools/clang/lib/Frontend/FrontendAction.cpp:901:0
#34 0x0000561fb7f438df clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) /root/checkedc/llvm/tools/clang/lib/Frontend/CompilerInstance.cpp:992:0
#35 0x0000561fb80e4c7e clang::ExecuteCompilerInvocation(clang::CompilerInstance*) /root/checkedc/llvm/tools/clang/lib/FrontendTool/ExecuteCompilerInvocation.cpp:252:0
#36 0x0000561fb626d73b cc1_main(llvm::ArrayRef<char const*>, char const*, void*) /root/checkedc/llvm/tools/clang/tools/driver/cc1_main.cpp:221:0
#37 0x0000561fb6262fef ExecuteCC1Tool(llvm::ArrayRef<char const*>, llvm::StringRef) /root/checkedc/llvm/tools/clang/tools/driver/driver.cpp:309:0
#38 0x0000561fb6263c16 main /root/checkedc/llvm/tools/clang/tools/driver/driver.cpp:388:0
#39 0x00007f4dcdcc8561 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x20561)
#40 0x0000561fb626086a _start (/opt/checkedc-llvm/bin/clang-5.0+0x189986a)
Stack dump:
0.	Program arguments: /opt/checkedc-llvm/bin/clang-5.0 -cc1 -triple x86_64-unknown-linux-gnu -emit-obj -mrelax-all -disable-free -main-file-name checkedc_bug.c -mrelocation-model static -mthread-model posix -mdisable-fp-elim -fmath-errno -masm-verbose -mconstructor-aliases -munwind-tables -fuse-init-array -target-cpu x86-64 -dwarf-column-info -debugger-tuning=gdb -resource-dir /opt/checkedc-llvm/lib/clang/6.0.0 -internal-isystem /usr/local/include -internal-isystem /opt/checkedc-llvm/lib/clang/6.0.0/include -internal-externc-isystem /usr/include/x86_64-linux-gnu -internal-externc-isystem /include -internal-externc-isystem /usr/include -fdebug-compilation-dir /root -ferror-limit 19 -fmessage-length 0 -fcheckedc-extension -fobjc-runtime=gcc -fdiagnostics-show-option -o /tmp/checkedc_bug-3c75f1.o -x c checkedc_bug.c 
1.	<eof> parser at end of file
2.	checkedc_bug.c:6:5: LLVM IR generation of declaration 'testfn'
3.	checkedc_bug.c:6:5: Generating code for declaration 'testfn'
clang-5.0: error: unable to execute command: Aborted
clang-5.0: error: clang frontend command failed due to signal (use -v to see invocation)
clang version 6.0.0 (https://github.com/Microsoft/checkedc-clang 48bd8fda1e5d1d929a64e9ad7b932bc6d9399591) (https://github.com/Microsoft/checkedc-llvm 9bf0f4a1bf12d8f7bd6259309b9352e3e93a396b)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/checkedc-llvm/bin
clang-5.0: note: diagnostic msg: PLEASE submit a bug report to https://github.com/Microsoft/checkedc-clang/issues and include the crash backtrace, preprocessed source, and associated run script.
clang-5.0: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang-5.0: note: diagnostic msg: /tmp/checkedc_bug-0fdae8.c
clang-5.0: note: diagnostic msg: /tmp/checkedc_bug-0fdae8.sh
clang-5.0: note: diagnostic msg: 

********************
```

I uploaded the two files which should be attached to the bug report to gist.github.com:

* https://gist.github.com/nmeum/001e0fa7291d5b06d6c47b1386646748
* https://gist.github.com/nmeum/a0f7dabc70efa691bc46a22aa5938ca5

As I have no experience with the LLVM code base I didn't dig any deeper. I hope that this bug report is useful to you none the less. In case you need additional information to reproduce the problem just let me know.

**Comment by dtarditi** [comment]
Thank you for the bug report.  The code snippet and the stack trace are plenty for us to go on.    We'll take a look at it  today.

**Comment by dtarditi** [comment]
The code generation code assumes that LValueBitCast expressions are always explicit expressions in the source code. For Checked C, we sometimes create LvalueBitCast expressions that are implicit expressions, when dealing with bounds-safe interfaces.  This should be pretty straightforward to fix.

**Comment by dtarditi** [comment]
This is fixed by PR #438, with a test added in PR https://github.com/Microsoft/checkedc/pull/245.  

We should add more tests to make sure other cases work (such as assignments to members with bounds-safe interfaces), but this should fix your specific issue.  I've opened https://github.com/Microsoft/checkedc/issues/246 to track adding the additional tests.

Let me know if you run into other issues
